rs-matter/src/transport/core.rs

```diff
@@ -103,680 +222,890 @@ impl<'a> Matter<'a> {
+                        packet,
+                        Some(&mut session),
+                        None,
+                        session_mgr.epoch,
+                        |wb| sc_write(wb, SCStatusCodes::CloseSession, &[]),
+                    )?;
+                }
 
-                    info!("Transport: exchange started");
+                Self::netw_send(send, packet.peer, &packet.buf[packet.payload_start..], true)
+                    .await?;
+            }
+            Err(e) => {
+                error!("\n>>>>> {packet}\n => Error ({e:?}), dropping");
+            }
+            Ok(new_exchange) => {
+                let meta = MessageMeta::from(&packet.header.proto);
+
+                if meta.is_standalone_ack() {
+                    // No need to propagate this further
+                    info!("\n>>>>> {packet}\n => Standalone Ack, dropping");
+                } else if meta.is_sc_status()
+                    && matches!(
+                        Self::is_close_session(&mut packet.buf[packet.payload_start..]),
+                        Ok(true)
+                    )
+                {
+                    warn!("\n>>>>> {packet}\n => Close session received, removing this session");
+
+                    let mut session_mgr = self.session_mgr.borrow_mut();
+                    if let Some(session_id) = session_mgr
+                        .get_for_rx(&packet.peer, &packet.header.plain)
+                        .map(|sess| sess.id)
+                    {
+                        session_mgr.remove(session_id);
+                    }
+                } else {
+                    info!(
+                        "\n>>>>> {packet}\n => Processing{}",
+                        if new_exchange { " (new exchange)" } else { "" }
+                    );
+
+                    debug!(
+                        "{}",
+                        Packet::<0>::display_payload(
+                            &packet.header.proto,
+                            &packet.buf[core::cmp::min(packet.payload_start, packet.buf.len())..]
+                        )
+                    );
+
+                    return Ok(true);
                 }
             }
         }
 
-        #[allow(unreachable_code)]
-        Ok::<_, Error>(())
+        Ok(false)
     }
 
-    #[inline(always)]
-    pub async fn exchange_handler<const N: usize, H>(
-        &self,
-        tx_buf: &mut [u8; MAX_TX_BUF_SIZE],
-        rx_buf: &mut [u8; MAX_RX_BUF_SIZE],
-        sx_buf: &mut [u8; MAX_RX_STATUS_BUF_SIZE],
-        handler_id: impl core::fmt::Display,
-        channel: &Channel<NoopRawMutex, ExchangeCtr<'_>, N>,
-        handler: &H,
-    ) -> Result<(), Error>
-    where
-        H: DataModelHandler,
-    {
-        loop {
-            let exchange_ctr: ExchangeCtr<'_> = channel.receive().await;
+    fn handle_accept_timeout_rx_packet<const N: usize>(&self, packet: &mut Packet<N>) -> bool {
+        if !packet.buf.is_empty() {
```

Comment:
I would try to use some early return here as well to reduce nesting and make things feel more linear.

Reply:
Comment seems to apply to below as well.